File: LPtHW/ex42.py
## animal is-a object (yes, sort of confusing) look at the extra credit
class Animal(object):

    def __init__(self):
        self.isAlive = True

## Dog is an Animal
class Dog(Animal):

    def __init__(self, name):
        ## Dog has-a name
        self.name = name

## Cat is an animal
class Cat(Animal):

    def __init__(self, name):
        ## Cat has-a name
        self.name = name

## Person is an object
class Person(object):

    def __init__(self, name):
        ## Person has-a name
        self.name = name

        ## Person has-a pet of some kind
        self.pet = None

## Employee is-a Person
class Employee(Person):

    def __init__(self, name, salary):
        ## In a class hierarchy with single inheritance, super can be used
        # to refer to parent classes without naming them explicitly, 
        # thus making the code more maintainable. This use closely parallels 
        # the use of super in other programming languages.
        # https://docs.python.org/2/library/functions.html#super
        super(Employee, self).__init__(name)
        ## Employee has-a salary
        self.salary = salary

## Fish is an object
class Fish(Animal):

    def __init__(self):
        super(Fish, self).__init__()
        self.swim = "It swims deftly through the water."

## Salmon is-a fish
class Salmon(Fish):
    
     def __init__(self):
         super(Salmon, self).__init__()
         self.isCanned = True
         self.isAlive = False

## Halibut is-a fish
class Halibut(Fish):

    def __init__(self):
        super(Halibut, self).__init__()
        self.beingEaten = True
        self.isAlive = "Not quite dead yet..."


## rover is-a Dog
rover = Dog("Rover")

## satan is-a Cat
satan = Cat("Satan")

## mary is-a Person
mary = Person("Mary")

## mary has-a pet that is-a Cat named satan
mary.pet = satan

## frank is an Employee
frank = Employee("Frank", 120000)

## frank has a pet that is-a Dog named rover
frank.pet = rover

## flipper is-a Fish (this flipper is not a dolphin. Dolphins are not fish.)
flipper = Fish()

## crouse is-a Salmon
crouse = Salmon()

## harry is-a Halibut
harry = Halibut()

print "Harry is Alive? ", harry.isAlive

print "Flipper swims? ", flipper.swim

print "Crouse is canned? ", crouse.isCanned
print "Crouse is Alive? ", crouse.isAlive

print "Harry is being eaten? ", harry.beingEaten
print "Harry is Alive? ", harry.isAlive

File: Learn_Ruby_the_Hard_Way/ex42/ex42.rb
## Animal is-a object look at the extra credit
class Animal
end

## Dog is-an Animal
class Dog < Animal

  def initialize(name)
    ## A dog has-a name
    @name = name
  end
end

## Cat is-an Animal
class Cat < Animal

  def initialize(name)
    ## a Cat has-a name
    @name = name
  end
end

## Person is-an Object (Not an Animal (In this implementation))
class Person

  def initialize(name)
    ## A Person has-a name
    @name = name

    ## Person has-a pet of some kind too
    @pet = nil
  end

  # allows pet to be read and written to
  attr_accessor :pet
end

## An Employee is-a Person
class Employee < Person

  def initialize(name, salary)
    ## Allows Employee to use its parent's constructor
    super(name)
    ## An Employee is a special kind of Person that has-a salary
    @salary = salary
  end

end

## A Fish is-an Object
class Fish
end

## A Salmon is-a Fish
class Salmon < Fish
end

## A Halibut is-a Fish
class Halibut < Fish
end

## rover is-a dog instance
rover = Dog.new("Rover")

## satan is-a Cat instance
satan = Cat.new("Satan")

## Mary is-a person insance
mary = Person.new("Mary")

## Mary has-a pet atrtribute, which is set to satan here
mary.pet = satan

## Frank is-an Employee instance
frank = Employee.new("Frank", 120000)

## Frank has-a pet attribute, which is set to rover here
frank.pet = rover

## Flipper is a Fish instance 
#  (Not a dolphin instance, which is-a Mammal, not a Fish)
flipper = Fish.new()

## Crouse is a Salmon instance
crouse = Salmon.new()

## Harry is a Halibut instance
harry = Halibut.new()
